### ISSUE TITLE
[HOLD for payment 2022-12-15] [$500] Login - Clicking the eye icon moves the cursor to the beginning

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#12018]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to URL https://staging.new.expensify.com/
2. Enter an email address
3. Press continue
4. Enter the password
5. Click to the eye icon

## Expected Result:
The cursor should not moved

## Actual Result:
The cursor moved to the beginning

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.18.2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/196955840-36e347cf-f392-43c0-8b63-822710a9d4d6.mp4

@Julesssss removed the capture as it was an unrelated recording

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Julesssss** [comment]
Yep, I can reproduce. @kbecciv would you mind re-uploading a video? I think you uploaded a different bug, so I removed it.

**Comment by Julesssss** [comment]
Oh, thanks for linking! While I agree with most of that discussion we've recently decided to try fixing each and every bug we encounter. For that reason, I think we should reopen this and see what sort of proposals we receive. 

If the solution is very convoluted, then maybe it would be preferable to just default to the end, given that mobile and some web browsers do this by default.

**Comment by kbecciv** [comment]
@Julesssss Video is attached, sorry for missing it.

**Comment by Julesssss** [comment]
Thanks, no worries

**Comment by Julesssss** [comment]
Awaiting proposals

**Comment by flaviadefaria** [comment]
link to the internal job posting: postinghttps://www.upwork.com/ab/applicants/1584804132817670144/job-details
 link external job: https://www.upwork.com/jobs/~017777eb694223c27d

**Comment by Someone-here** [comment]
Apparently the cursor resetting is just react-native internal behaviour, I tested on snack and got the cursor going to the beginning as well. I tried to fiddle around with other textInput components which was unsuccessful but I think that would be unnecessary and just add extra code. This was the solution that was the simplest and most importantly added the least amount of code. And is much more reliable than relying on a third-party component.

**Comment by Julesssss** [comment]
I agree we don't want to depend on 3rd party component for this. I'm coming around to the manual setting of cursor, but want to leave the issue open a little longer.

**Comment by Julesssss** [comment]
Holding for any last proposals

**Comment by flaviadefaria** [comment]
Posted in [slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1667844542973769) re next steps.

**Comment by Someone-here** [comment]
Hey @Julesssss now could we move ahead with that solution?

**Comment by Julesssss** [comment]
Yep, seems like this is the best solution we have.

**Comment by Julesssss** [comment]
Oh, thanks for mentioning @aimane-chnaif. Would you mind explaining further?

As far as I can tell the proposal would only set the cursor on toggle, but maybe I'm missing something?

**Comment by Julesssss** [comment]
Actually @Someone-here, sorry to halt hire this but there are a couple of issues: 

> and will also do some extensive research on all possible solutions to this.

1) We need to approve the proposal before sending you off to investigate. Ideally the investigation occurs beforehand, so that we can properly compare solutions. 

2) I missed this previously, but the video shows a delay in the cursor being set (at 11-12 seconds). I'd say it's better to have the current bug than to show a laggy movement of the cursor. 

3) As @aimane-chnaif mentions, the solution relies on setTimeout AND we lose the current position of the cursor. 

I'm going to unassign for now, but we're open to any alternative solutions.

**Comment by Julesssss** [comment]
@flaviadefaria would you mind doubling the bounty? The solution will require more thought.

**Comment by Someone-here** [comment]
I found a way to remove the setTimeOut as well! We can just pass the function as an argument to the setState, which will not result in that delay

**Comment by Someone-here** [comment]
Just let me think of a proper solution to preserving the cursor position and I'll write another proposal.

**Comment by flaviadefaria** [comment]
>@flaviadefaria would you mind doubling the bounty? The solution will require more thought.

@Julesssss done!

**Comment by Someone-here** [comment]
I made it so the position of the cursor stays the same as well by using the existing this.state.selection

https://user-images.githubusercontent.com/77237602/200622247-f2e0d49b-238f-465f-9574-7fbf9cd788e4.mp4

Now the only thing left is removing the setTimeout, turns out that the setState is for some reason not calling the callback function.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by Someone-here** [comment]
Is it essential to remove setTimeOut from the solution?

**Comment by Someone-here** [comment]
Ok, I removed setTimeOut, and even removed the manual selection. Turns out that if we change the state for the password right in the onMouse down rather than onPress, that solves the problem on web.

![check](https://user-images.githubusercontent.com/77237602/200887955-8f3a8594-7518-4b98-9a0a-4c78532adb45.png)

https://user-images.githubusercontent.com/77237602/200888208-2e8bc1ed-d701-4b99-bc45-aa6b6fdbf24e.mp4

Now only left to figure it out on Android and ios. Actually I have a windows machine and cannot get the Android app to run, this might not be the right place to ask but could you please tell me the steps to run it on windows? I have WSL also if that's needed.

**Comment by eVoloshchak** [comment]
> Actually I have a windows machine and cannot get the Android app to run, this might not be the right place to ask but could you please tell me the steps to run it on windows? I have WSL also if that's needed.

In my personal experience, it will be much faster to install linux as a second system, windows is _**tricky**_

**Comment by flaviadefaria** [comment]
Based on this comment I'll try to reproduce this again to see if it's still an issue.
>Issue not reproducible during KI retests. (First week)

**Comment by Someone-here** [comment]
This is fixed on chrome, in the last comment by me. Since I don't own a macbook would someone like to collaborate with me to find a solution for iOS as well?

**Comment by eVoloshchak** [comment]
> This is fixed on chrome, in the last comment by me. Since I don't own a macbook would someone like to collaborate with me to find a solution for iOS as well?

I can test your branch on iOS if you already have a solution
If not, I can recommend using macOS instances, we have a handy guide: [TESTING_MACOS_AND_IOS.md](https://github.com/Expensify/App/blob/main/contributingGuides/TESTING_MACOS_AND_IOS.md)

**Comment by Someone-here** [comment]
Okay, I tested on ios using snack https://snack.expo.dev/CPp4q2OGd and it is just native behavior for the cursor to go to the end. Even if we want to fix this, we would need a work-around from this which would require having a separate selection state and also setTimeOut because there is no react lifecycle hook that can update the selection state predictably after the component updates. 

If we decide to use setTimeOut, then this can be fixed on iOS. On chrome it works just by toggling the secure state in the onMouseDown event instead of onPress because onPress is designed for mobile platforms where as onMouseDown is for web.

So how should I proceed now? In my opinion, we should just fix this on chrome and on iOS, the cursor is still at the end not the beginning, so it wouldn't be worth having a workaround for just preserving the cursor state when in between the text. But it can be any way you require. @Julesssss

**Comment by eVoloshchak** [comment]
> for chrome we can streamline it with android just by calling `this.togglePassword` in the `onMouseDown` event instead of `onPress` event

@Someone-here, I think that would break the accessibility, you wouldn't be able to toggle password visibility using keyboard

**Comment by Someone-here** [comment]
@eVoloshchak, I fixed that too, I added both events `onPress` and `onMouseDown` and now it works on all platforms and can even be changed by keyboard also

**Comment by eVoloshchak** [comment]
@Someone-here, that's great!
Could you please post an updated proposal?

**Comment by Someone-here** [comment]
### PROPOSAL

**Option 1: (Web)**
We can call the `this.togglePassword` method in the `onMouseDown` event as well as `onPress` event. This will preserve the cursor position on all devices except iOS.

![● BaseTextInput js - app  WSL_ Ubuntu  - Visual Studio Code 15-11-2022 00_18_08](https://user-images.githubusercontent.com/77237602/201741646-50a5b334-7cb8-4c03-87a3-896d5176ae27.png)

In iOS, the native behavior is to set the cursor to the end to the input even when the `e.preventDefault` method is used. But this is very minor as the cursor doesn't even reset to the start, it stays at the end. Only if you keep the cursor in the middle will you notice the difference.

**Option 2: (Web, iOS)**
We can overcome the native behavior by keeping track of the selection in a special state that can be used to update the component after click. Since there is no react lifecycle hook that can precisely execute the `setSelection` when the `TextInput`
updates, we would wither have to create a wrapper component specially for that or use `setTimeOut`.

Here is the result after using Option 1:

https://user-images.githubusercontent.com/77237602/201742108-c7e978fb-74e7-4924-83fe-86e0e57c2a40.mp4

**Comment by eVoloshchak** [comment]
@Someone-here's proposal looks good to me. But I still feel like we should try to achieve the same behavior on different platforms. We try to avoid using `setTimeOut` unless absolutely necessary, is there another way to make it work?

**Comment by Someone-here** [comment]
I tried all the lifecycle hooks and component events, unfortunately none of them execute after `secureEntryText` is changed. So in order to execute a function after that, we will have to use `setTimeout`. But it very reliable and not very noticeable as the delay would be set to 0ms.

**Comment by eVoloshchak** [comment]
@Julesssss, we have a proposal that streamlines behavior on all platforms except iOS. There it uses the native behavior of setting the cursor to the end. It seems like it's impossible to overcome this without using `setTimeount`, but still, should we wait for more proposals?

**Comment by Julesssss** [comment]
I'm leaning towards option1 for now, then treating iOS as a separate, lower-priority issue. It doesn't seem like we have any good alternatives here

**Comment by Someone-here** [comment]
Yes, I think Option 1 is better as well. Better to keep it at a minimal solution

**Comment by eVoloshchak** [comment]
> I'm leaning towards option1 for now, then treating iOS as a separate, lower-priority issue. It doesn't seem like we have any good alternatives here

In that case, let's go with @Someone-here's [proposal](https://github.com/Expensify/App/issues/12018#issuecomment-1314225656) (option 1)
🎀👀🎀 C+ reviewed!

**Comment by flaviadefaria** [comment]
Waiting for PR to be merged.

**Comment by Someone-here** [comment]
Made the PR: https://github.com/Expensify/App/pull/12941
Please review

**Comment by Someone-here** [comment]
@eVoloshchak or @Julesssss Could you please test this proposal on iOS and Safari? I don't own a mac currently, therefore it would mean a lot to me if you would do so...

**Comment by Julesssss** [comment]
Reviewing the PR today

**Comment by Someone-here** [comment]
@eVoloshchak @Julesssss Sorry for the inconvenience but I closed the PR by mistake while trying to sync the repo with main. Let me just open a new PR for that

**Comment by Someone-here** [comment]
Here is the new PR: https://github.com/Expensify/App/pull/13103
Please review.
I hope this didn't cause much of an issue.

**Comment by flaviadefaria** [comment]
Bump @eVoloshchak @Julesssss!

**Comment by eVoloshchak** [comment]
Reviewing

**Comment by Julesssss** [comment]
Seems like this introduced this regression, so I'm going to have to revert it.

**Comment by Someone-here** [comment]
I'm really sorry for this. I'll fix it asap

**Comment by Julesssss** [comment]
No worries. I also missed this.

**Comment by Someone-here** [comment]
I found out a solution. Should I make a new PR?

**Comment by Julesssss** [comment]
> I found out a solution. Should I make a new PR?

Yeah, just give me a few minutes to revert the previous changes to unblock our staging build. It'll be a few more minutes, and then you can raise the PR against a clean `main` branch.

**Comment by Julesssss** [comment]
@Someone-here all done, no rush with the new PR as I won't be able to review until next week

**Comment by Someone-here** [comment]
Okay, thank you for giving me another chance. I will do extensive testing before submitting PR this time.

**Comment by Someone-here** [comment]
Made a new PR https://github.com/Expensify/App/pull/13317 This fixes the regression. I have done extensive testing with checkboxes on all platforms as well. I hope this serves its purpose this time...

**Comment by flaviadefaria** [comment]
Heads up I'm heading OoO for a week so in case this needs to be paid out before I come back then please reassign or ask someone else in the team to process the payment. Thanks!

**Comment by Julesssss** [comment]
I decided to close the outstanding PR on the second review as it's adding quite a bit of complexity to the Checkbox component. Explained a bit more [here](https://github.com/Expensify/App/pull/13317).

@flaviadefaria could you please pay out the original contract on Monday? We had a regression so I believe it will be 50% of the payment

**Comment by eVoloshchak** [comment]
It also was merged within ~8 days, which is a 50% penalty

I don't think I'm eligible for any payment on this one, there were a couple of regressions I could have noticed, we've dragged this out for more than 2 weeks, I don't want to get paid for this one

**Comment by Julesssss** [comment]
Fair enough, thanks @eVoloshchak but you still reviewed the PR originally so I think the reduced payment is fine. 

Honestly, I'm not sure the 12-day rate counts here due to the regression.

**Comment by Someone-here** [comment]
Guys, I am really sorry if you think I did not do upto to the mark but I tried my best....

**Comment by Julesssss** [comment]
Hey @Someone-here don't worry. I should have called it out earlier in the process and maybe we could have discussed alternatives. Let's just move onto the next issue 👍

**Comment by Someone-here** [comment]
Great! Thank you for the positive attitude

**Comment by flaviadefaria** [comment]
We'll be paying out $250 for each of you (50% penalty). I've sent you both the contract offer in UW please accept and let me know here once done. Thanks!

**Comment by flaviadefaria** [comment]
Everyone has been paid so closing this out.
